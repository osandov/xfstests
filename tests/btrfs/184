#! /bin/bash
# SPDX-License-Identifier: GPL-2.0
# Copyright (c) 2019 Omar Sandoval.  All Rights Reserved.
#
# FS QA Test 184
#
# Create files with preallocated extents in various patterns and send them with
# btrfs send --stream-version 2. Test for "Btrfs: send, use fallocate for
# preallocated extents and holes".
#
seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"

here=`pwd`
tmp=/tmp/$$
status=1	# failure is the default!
trap "_cleanup; exit \$status" 0 1 2 3 15

_cleanup()
{
	cd /
	rm -f $tmp.*
}

# get standard environment, filters and checks
. ./common/rc
. ./common/filter
. ./common/punch

# remove previous $seqres.full before test
rm -f $seqres.full

_supported_fs btrfs
_supported_os Linux
_require_scratch
_require_xfs_io_command fiemap
_require_xfs_io_command fpunch
_require_btrfs_send_stream_version 2

_scratch_mkfs >> $seqres.full 2>&1
_scratch_mount

_run_btrfs_util_prog subvolume create "$SCRATCH_MNT/subvol"

$XFS_IO_PROG -f -c "falloc 0 $((128 * 512))" "$SCRATCH_MNT/subvol/prealloc"

$XFS_IO_PROG -f -c "truncate $((128 * 512))" \
	-c "falloc -k 0 $((32 * 512))" \
	-c "falloc -k $((64 * 512)) $((32 * 512))" \
	"$SCRATCH_MNT/subvol/prealloc_with_holes"

$XFS_IO_PROG -f -c "truncate $((128 * 512))" \
	-c "falloc -k $((128 * 512)) $((32 * 512))" \
	-c "falloc -k $((256 * 512)) $((32 * 512))" \
	"$SCRATCH_MNT/subvol/prealloc_beyond_eof"

$XFS_IO_PROG -f -c "truncate $((128 * 512))" \
	-c "falloc -k $((96 * 512)) $((64 * 512))" \
	"$SCRATCH_MNT/subvol/prealloc_straddles_eof"

$XFS_IO_PROG -f -c "pwrite 0 $((128 * 512))" \
	-c "fpunch $((64 * 512)) $((32 * 512))" \
	-c "falloc -k $((128 * 512)) $((32 * 512))" \
	-c "falloc -k $((192 * 512)) $((32 * 512))" \
	"$SCRATCH_MNT/subvol/data_and_hole_and_prealloc" | _filter_xfs_io

_run_btrfs_util_prog subvolume snapshot -r "$SCRATCH_MNT/subvol" "$SCRATCH_MNT/snap"
_run_btrfs_util_prog send --stream-version 2 -f "$tmp.send" "$SCRATCH_MNT/snap"

_scratch_unmount
_check_scratch_fs

_scratch_mkfs >> $seqres.full 2>&1
_scratch_mount

_run_btrfs_util_prog receive -f "$tmp.send" "$SCRATCH_MNT"

md5sum "$SCRATCH_MNT/snap/prealloc" | _filter_scratch
$XFS_IO_PROG -r -c "fiemap -v" "$SCRATCH_MNT/snap/prealloc" | _filter_fiemap
md5sum "$SCRATCH_MNT/snap/prealloc_with_holes" | _filter_scratch
$XFS_IO_PROG -r -c "fiemap -v" "$SCRATCH_MNT/snap/prealloc_with_holes" \
	| _filter_fiemap
md5sum "$SCRATCH_MNT/snap/prealloc_beyond_eof" | _filter_scratch
$XFS_IO_PROG -r -c "fiemap -v" "$SCRATCH_MNT/snap/prealloc_beyond_eof" \
	| _filter_fiemap
md5sum "$SCRATCH_MNT/snap/prealloc_straddles_eof" | _filter_scratch
$XFS_IO_PROG -r -c "fiemap -v" "$SCRATCH_MNT/snap/prealloc_straddles_eof" \
	| _filter_fiemap
md5sum "$SCRATCH_MNT/snap/data_and_hole_and_prealloc" | _filter_scratch
$XFS_IO_PROG -r -c "fiemap -v" "$SCRATCH_MNT/snap/data_and_hole_and_prealloc" \
	| _filter_fiemap

status=0
exit
