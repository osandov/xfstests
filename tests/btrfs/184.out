QA output created by 184
wrote 65536/65536 bytes at offset 0
XXX Bytes, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
fcd6bcb56c1689fcef28b57c22475bad  SCRATCH_MNT/snap/prealloc
0: [0..127]: unwritten
fcd6bcb56c1689fcef28b57c22475bad  SCRATCH_MNT/snap/prealloc_with_holes
0: [0..31]: unwritten
1: [32..63]: hole
2: [64..95]: unwritten
3: [96..127]: hole
fcd6bcb56c1689fcef28b57c22475bad  SCRATCH_MNT/snap/prealloc_beyond_eof
0: [0..127]: hole
1: [128..159]: unwritten
2: [160..255]: hole
3: [256..287]: unwritten
fcd6bcb56c1689fcef28b57c22475bad  SCRATCH_MNT/snap/prealloc_straddles_eof
0: [0..95]: hole
1: [96..159]: unwritten
0519b05a43ef6fd071196d0d5ff07779  SCRATCH_MNT/snap/data_and_hole_and_prealloc
0: [0..63]: data
1: [64..95]: hole
2: [96..127]: data
3: [128..159]: unwritten
4: [160..191]: hole
5: [192..223]: unwritten
